### ISSUE TITLE
Torch script to support dictionary with keys of type tuple

### ISSUE THREAD

**Issue by neginraoof** [pytorch/pytorch#39356]
## 🚀 Feature
I'm seeing the following error in torch script when scripting an nn.Module that contains a dictionary of tuple (or list) data type as keys.

RuntimeError: Cannot create dict for key type '(int)', only int, float, Tensor and string keys are supported.

I'm looking into scripting a model that uses dictionary of tuple keys (Dict[Tuple[int], float])
to store tuples of ngrams with the corresponding ngram probability.

## Pitch
I expect torch script to successfully script the defined module with an attribute of type 
(Dict[Tuple[int], float]).

## Additional context

Repro code:

```python
# Scripting the module below throws the above exception:
import torch
from torch.jit.annotations import Tuple, List, Dict

class MyModule(torch.nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.dict = torch.jit.Attribute({}, Dict[Tuple[int], float])

    def forward(self, x):
        return self.dict.get(x)

torch.jit.script(MyModule())
```

**Comment by eellison** [comment]
Is there anything preventing unpacking the tuple here ? `Dict[int, float]`

**Comment by neginraoof** [comment]
Yes, there is a specific use-case for dictionaries of tuple keys. An example is a language model where keeps track of ngrams and log probabilities corresponding to each ngram. In this case, ngrams should be represented with a tuple/list.

**Comment by eellison** [comment]
`List[int]` aren't hashable in python, so it sounds like the request is for keys of a Tuple of arbitrary length. We can't do that because Tuples in TorchScript are of fixed length. What could do here is add support for hashing a custom object that defines a `__hash__` function.

**Comment by eellison** [comment]
You can workaround this by writing your own hash function.

```
class MyModule(torch.nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.dict = torch.jit.Attribute({}, Dict[int, Tuple[List[int], float]])


    def hash(self, x: List[int]):
        # dumb hash function
        s = ""
        for elem in x:
            s += str(elem)
        return hash(s)

    @torch.jit.export
    def set_item(self, key: List[int], value: float):
        key_hash = self.hash(key)
        self.dict[key_hash] = (key, value)

    def forward(self, x: List[int]):
        key = self.hash(x)
        return self.dict.get(key)
```

**Comment by neginraoof** [comment]
@eellison Thanks for the suggestion. How about keys of tuple of fixed length? Actually, in the case of this model, tuples are of fixed lenght. How is that specified in the annotation if I want to test it?

**Comment by neginraoof** [comment]
@eellison @bzinodev 

For the current model we're looking to export, we are able to use the work around to avoid Dict[Tuple, floa] with relatively small user code changes.

**Comment by eellison** [comment]
@neginraoof glad that worked for you! I should amend my answer to use the str as the key, and not the `hash(str)`, so that way hash collisions are correctly handled.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
